doc/configuration.md

```diff
@@ -0,0 +1,282 @@
+   - If the `--config` parameter points to a directory, `azure-init` follows this hierarchy:
+     - First, it looks for a base file named `azure-init.toml` in the directory.
+     - Then, it merges any additional `.toml` files in a `.d` subdirectory within the specified directory, in lexicographical order.
+     ```text
+     /etc/azure-init/
+     ├── azure-init.toml            # Base configuration
+     └── .d/
+         ├── 01-network.toml        # Additional network configuration
+         ├── 02-ssh.toml            # Additional SSH configuration
+         └── 99-overrides.toml      # Final overrides
+     ```
+
+#### 3. Defaults in Code
+   - If neither a file nor a directory is provided, `azure-init` falls back to using the default values specified in `config.rs`.
+
+### Example 1: Single Configuration File
+**Command:**
+```sh
+azure-init --config /path/to/custom-config.toml
+```
+**Order of Merging:**
+1. Loads `/path/to/custom-config.toml`.
+2. Applies any CLI overrides if present.
+3. Fills in missing values with defaults from `config.rs`.
+
+### Example 2: Directory with Multiple .toml Files
+
+**Command:**
+```sh
+azure-init --config /path/to/custom-config-directory
+```
+**Directory Structure**
+```bash
+/path/to/custom-config-directory/
+├── azure-init.toml                # Base configuration
+└── .d/
+    ├── 01-network.toml            # Network configuration
+    ├── 02-ssh.toml                # SSH configuration
+    └── 99-overrides.toml          # Overrides
+```
+**Order of Merging:**
+
+1. Loads `azure-init.toml` as the base configuration.
+2. Merges `.d` files in alphabetical order:
+   - `01-network.toml`
+   - `02-ssh.toml`
+   - `99-overrides.toml`
+3. Applies any CLI overrides if present.
+4. Fills in missing values with defaults from `config.rs`.
+
+### Example 3: Default Path without --config
+**Assumption:** If no `--config` path is specified, `azure-init` will only use the default values set within `config.rs`.
+
+**Order of Loading:**
+
+1. Loads configuration directly from defaults specified in `config.rs`.
+
+## Validation Behaviors
+
+Custom enum types for fields like `SshAuthorizedKeysPathQueryMode` validate that config settings set by the user match the supported options. If a user enters an unsupported config value, deserialization will fail because Serde will not be able to map that value to one of the enum variants.
+
+### Example: 
+```toml
+# Configure the authorized_keys_path_query_mode type.
+[ssh]
+authorized_keys_path_query_mode = "disabled"
+```
+
+## Configuration Fields
+
+#### Ssh Struct:
+- **authorized_keys_path**: `PathBuf`
+  - **Default**: `~/.ssh/authorized_keys`
+  - **Description**: Specifies the file path to the SSH authorized keys.
+- **configure_password_authentication**: `bool`
+  - **Default**: `false`
+  - **Description**: Controls whether password authentication is configured for SSH.
+- **authorized_keys_path_query_mode**: `SshAuthorizedKeysPathQueryMode`
```

Comment:
I think we only have two behaviors: querying using `sshd -G`, or not querying it. In that case this one should be a boolean?

Reply:
Agreed, I would suggest keeping the proposed authorized_keys_path and dropping the fallback_sshd_default option, e.g.

[sshd]
query_sshd_config: bool = true
authorized_keys_path: string = .ssh/authorized_keys

the scope for querying the sshd_config can be useful beyond authorizedkeys, e.g. PasswordAuthentication handling so I'd keep the name generic.

- if query_sshd_config = true and sshd -G succeeds, authorizedkeys is determined dynamically from sshd.

- if query_sshd_config = true and sshd -G fails, failure is reported but VM will still be configured according to authorized_keys_path, which likely is the default .ssh/authorized_keys (we still want to configure the VM as much as possible even if reporting failure)

- if query_sshd_config = false, authorized_keys_path is used as-is.